misk-crypto/src/main/kotlin/misk/crypto/CiphertextFormat.kt

```diff
@@ -0,0 +1,304 @@
+          }
+          if (value.size >= Short.MAX_VALUE) {
+            throw InvalidEncryptionContextException("value is too long")
+          }
+          buffer.put(encodeVarInt(key.size))
+          buffer.put(key)
+          buffer.put(encodeVarInt(value.size))
+          buffer.put(value)
+        }
+      } catch (e: BufferOverflowException) {
+        throw InvalidEncryptionContextException("encryption context is too long")
+      }
+      val aad = ByteArray(buffer.position())
+      buffer.flip()
+      buffer.get(aad)
+      return aad
+    }
+
+    @VisibleForTesting
+    internal fun deserializeEncryptionContext(aad: ByteArray?): Map<String, String>? {
+      if (aad == null) {
+        return null
+      }
+      val src = DataInputStream(ByteArrayInputStream(aad))
+      val entries = decodeVarInt(src)
+      if (entries == 0) {
+        return null
+      }
+      return (1..entries).map {
+        val keySize = decodeVarInt(src)
+        val keyBytes = ByteArray(keySize)
+        src.readFully(keyBytes)
+        val valueSize = decodeVarInt(src)
+        val valueBytes = ByteArray(valueSize)
+        src.readFully(valueBytes)
+        keyBytes.toString(Charsets.UTF_8) to valueBytes.toString(Charsets.UTF_8)
+      }.toMap()
+    }
+
+    private fun readCiphertext(src: DataInputStream) : ByteArray {
+      val ciphertextStream = ByteArrayOutputStream()
+      var readByte = src.read()
+      while(readByte >= 0) {
+        ciphertextStream.write(readByte)
+        readByte = src.read()
+      }
+      return ciphertextStream.toByteArray()
+    }
+
+    private const val SEPTET = (1 shl 7) -1
+    private const val HAS_MORE_BIT = 1 shl 7
+
+    private fun encodeVarInt(integer: Int): ByteArray {
+      val list = mutableListOf<Byte>()
+      var int = integer
+      var byte = int and SEPTET
+      while(int shr 7 > 0) {
+        list.add((byte or HAS_MORE_BIT).toByte())
+        int = int shr 7
+        byte = int and SEPTET
+      }
+      list.add(int.toByte())
+
+      return list.toByteArray()
+    }
+
+    private fun decodeVarInt(src: DataInputStream): Int {
+      var byte = src.readByte().toInt()
+      var integer = byte and SEPTET
+      while (byte and HAS_MORE_BIT > 0) {
+        byte = src.readByte().toInt()
+        integer += ((byte and SEPTET) shl 7)
+      }
+      return integer
+    }
+
+    /**
+     * Extracts the ciphertext and encryption context from the [serialized] ByteArray.
+     *
+     * This method is meant to be used with field-level-encryption in Hibernate only.
+     */
+    fun deserializeFleFormat(serialized: ByteArray): Pair<ByteArray, Map<String, String?>> {
+      val src = DataInputStream(ByteArrayInputStream(serialized))
+      val version = src.readByte().toInt()
+      if (version != 1) {
+        throw InvalidCiphertextFormatException("invalid version")
+      }
+      val bitmask = src.readInt()
+      if (bitmask > Short.MAX_VALUE) {
```

Comment:
is this guaranteed to not happen with the old fle format?

Reply:
Not exactly, for FLE ciphertext format, we allocate an `int` fo the bitmask, but then encode/decode it as a `short`.
But, it is technically possible to have a bit set outside of that scope.
The FLE format is also more loosely defined, so this is a good indicator of whether a serialized ciphertext is indeed encoded in that format.